Add allow origin header in error responses

bjoernricks opened the pull request:
Hi,

thanks for merging my last patch. This one adds the cors header to error responses too. After reading the cors W3C recommendation I am still not 100% sure if it is required for responses. However Firefox and Chrome are complaining about missing cors for 4xx responses.

ejurgensen commented:
Could you say a bit more about what the goal of this is?

bjoernricks commented:
Of course!

The final goal is to get at working [daap client](https://github.com/bjoernricks/daap.js) for forked-daapd in a browser. Currently that isn't possible because for http requests the CORS header Access-Control-Allow-Origin in the response is missing. Without the header the [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) will fail in all (current) browsers. My first merge request fixed that for 200 responses but the header is still missing for 4xx responses. This merge request tries to fix it.

ejurgensen commented:
Sorry, need a bit more elaboration: Since 4xx are failures anyway, why is the header required?

I see you have taken some of the libevent code and copied it into forked-daapd. I'm not so thrilled about that. At least you would need to adjust the coding style + add explanation and reference.

bjoernricks commented:
No problem. I'll try to give a bit more details about the cors implementations.

After reading the specs (http://www.w3.org/TR/cors/) several times I came to the conclusion that the header should always be set despite it's not explicitly mentioned. IMHO the Access-Control-Allow-Origin Header is required for all requests to urls that don't fall under the same origin rule. After checking the implementations in current Firefox and Chrome the following results can be observed:

* For 2XX responses without Access-Control-Allow-Origin headers the javascript handler isn't called at all. Even not any error handler.

* For 4XX responses an error handler is called but you also get a big fat error message about missing cors headers in the javascript console that can't be avoided. Also I haven't checked if it is possible to evaluate the response (status, message, headers, ..) at all yet.

Of course I can understand that you are not thrilled about the copied code. I wouldn't be that either too. I also thought about adding an issue report to libevent because removing all set headers can be wrong for many applications. But I am unsure wether the function is intended for a general error reply or only a small helper. Therefore adding a more or less own implementation seems to be more correct to me. That is also an advantage e.g. if the error page must be changed in future. Now it is only one function to change.

Which coding style is not correct? I have tried to use the same style as already used. Btw. the existing code has many mixed tabs and spaces indentation. Of course I can add some references to the cors spec and some explanations to the http_error_reply function.

ejurgensen commented:
I understand the point of the header, but my thinking was:
  4xx response with header: browser shows an error message (maybe?)
  4xx response w/o header: browser shows another error message
So what I am interested in knowing is the practical difference for the user of these two scenarios. It would be great if you could clarify that for me.

Another thing: If the option is not set, then I think http_send_error = evhttp_send_error, right? This should be reflected in http_send_error, so that the path just leads straight to evhttp_send_error when the option is not set.

Also, you changed some evhttp_send_error to http_send_error, but not others. Which ones are in scope, and which ones are out of scope? And why?

About coding style:
1. Avoid mid-function declarations (httpd.c line 1512)
2. All functions in httpd.h are prefixed with http**d**
3. forked-daapd's indentation is indeed a mix of tabs and spaces (which I'm not a huge fan of). But it is a consistent mix, so please use the same.
4. Note the placement and indentation of brackets

roblan commented:
I'm not sure but 4xx without CORS header could not give correct error information which can be important in some error handling cases (for example 403 vs 404).

bjoernricks commented:
> I'm not sure but 4xx without CORS header could not give correct error information which can be important in some error handling cases (for example 403 vs 404).

yes that is what i am expecting too.

bjoernricks commented:
Just run a test with Firefox and Chrome without adding CORS headers in error responses. Both browser behave the same by setting the responseText to "" and the status to 0. That mean I would not be able to detect a failed login in JavaScript without my patches.

ejurgensen commented:
Ok, thanks for the update. If you could fix the things I mentioned above I will include it.

bjoernricks commented:
> 1. Avoid mid-function declarations (httpd.c line 1512)

Do you mean I should use the global evbase_http variable here or only move the variable declaration to the top of the function?

> 2. All functions in httpd.h are prefixed with httpd

Ups I'll fix that.

> 3. forked-daapd's indentation is indeed a mix of tabs and spaces (which I'm not a huge fan of). But it is a consistent mix, so please use the same.

After looking deeply at src/httpd.c the indentation is first 2 spaces and afterwards tabs. Is that right?

> 4. Note the placement and indentation of brackets

I am going to fix that too.

bjoernricks commented:
> Also, you changed some evhttp_send_error to http_send_error, but not others. Which ones are in scope, and which ones are out of scope? And why?

Imho I did change all evhttp_send_error in the daap code paths. I wasn't sure if you would like to see changes in all your code.

> Another thing: If the option is not set, then I think http_send_error = evhttp_send_error, right? This should be reflected in http_send_error, so that the path just leads straight to evhttp_send_error when the option is not set.

I'll change my code accordingly.

ejurgensen commented on src/dmap_common.c:
```diff
@@ -355,7 +355,7 @@ dmap_send_error(struct evhttp_request *req, const char *container, const char *e
       DPRINTF(E_LOG, L_DMAP, "Could not allocate evbuffer for DMAP error\n");
 
-      evhttp_send_error(req, HTTP_SERVUNAVAIL, "Internal Server Error");
+      httpd_send_error(req, HTTP_SERVUNAVAIL, "Internal Server Error");
```
dmap_send_error uses evhttp_send_reply, which won't include the headers. So with the changes you have made here, dmap_send_error will only include the headers on these rare out of memory edge cases. I don't understand the point of that.

bjoernricks replied:
You are totally right. If httpd_send_errror is used the allow origin header should be added here too.

bjoernricks replied:
Regarding adding the headers at several places in the code how would you implement that?

Should I add a new function like httpd_add_cors_headers(struct evkeyvalq *headers) and call it at several places or better to have a function like httpd_send_reply_internal which is more or less a wrapper around evhttp_send_reply like in my httpd_send_error and replace calls to evhttp_send_reply with this new function?

ejurgensen commented on src/dmap_common.h:
```diff
@@ -6,6 +6,7 @@
 #include <event2/http.h>
 
 #include "db.h"
+#include "httpd.h"
```
Does not belong here, is not referenced in dmap_common.h. Should be in the .c - if it is required at all.

bjoernricks replied:
Of course!

ejurgensen commented on src/httpd.c:
```diff
@@ -1485,3 +1491,39 @@ httpd_deinit(void)
 }
+
+void
+httpd_send_error(struct evhttp_request* req, int error, const char* reason)
```
Please add explanation and reference to the original libevent source. Also: The function is misplaced here at the end of the file, it should of course come after httpd_send_reply.

ejurgensen commented on src/httpd.h:
```diff
@@ -23,4 +23,7 @@ httpd_init(void);
 httpd_deinit(void);
 
+void
+httpd_send_error(struct evhttp_request* req, int error, const char* reason);
```
Also misplaced

ejurgensen commented:
I've added a few comments to the proposed changes.

> Imho I did change all evhttp_send_error in the daap code paths. I wasn't sure if you would like to see changes in all your code.

What I was thinking is that you changed httpd_daap.c, but not httpd_dacp.c for instance. I guess that makes sense if your client is strictly daap. But what about streaming the files, which is in httpd.c? How come you don't need the header there?

Even in httpd_daap.c there are still a few calls to evhttp_send_reply(), and they also won't have the header, right?

bjoernricks commented:
I have fixed most of your proposed changes.

> What I was thinking is that you changed httpd_daap.c, but not httpd_dacp.c for instance. I guess that makes sense if your client is strictly daap. 

My intention was to only change a minimal set of functions. Therefore I did only touch the daap code paths. Imho it would be nice to add the cors header to the dacp code too but I haven't looked at implementing dacp yet.

> But what about streaming the files, which is in httpd.c? How come you don't need the header there?

I am currently not sure about that but I am going to test it.

bjoernricks commented:
>  > But what about streaming the files, which is in httpd.c? How come you don't need the header there?
> 
> I am currently not sure about that but I am going to test it.

The cors header is not necessary for using it as the <audio> source.

> Even in httpd_daap.c there are still a few calls to evhttp_send_reply(), and they also won't have the header, right? 

Yes indeed. I even missed a very important function: httpd_basic_auth. I will fix it after we decided how to add the headers at several places. 

Btw. it is really cool to be able to stream audio from forked-daapd directly to the browser now! Thanks for caring about my patches!

bjoernricks commented:
> Yes indeed. I even missed a very important function: httpd_basic_auth. I will fix it after we decided how to add the headers at several places.

I have thought about this problem for a while and now I added my proposal in commit c0efb9434ff29254e5dd802eb31073548d28d609.

Sadly during testing the authentication I found out that I am missing the "preflight" case. Initially I thought  the "preflight" case is only necessary for "custom" headers and "custom" header meant a non standardized header and therefore not the authorization header. But that's not the case. After adding the authorization header in javascript the "preflight" path is required. For details see http://www.html5rocks.com/en/tutorials/cors/

Locally I have already implemented the preflight code path and if you are fine with my latest patches I'll push them too. But before I need to know which http methods should be allowed in forked-daapd. Are there any code paths which require other methods then GET?
By default (and that's currently implemented in forked-daapd) GET, POST, HEAD, PUT and DELETE are allowed in libevent (libevent/http.c evhttp_new_object). I would like to change that to GET and OPTIONS only.

ejurgensen commented:
I think that instead of inserting httpd_handle_cors_simple() in multiple locations, it would be better to change all occurances of evhttp_send_reply with httpd_send_reply, and then make the gzipping that httpd_send_reply does optional for the caller. Otherwise it becomes very inconsistent what headers forked-daapd returns.

It might be better if I do this myself, since I have an idea of how I want this to be.

BTW I don't think httpd_basic_auth is actually ever invoked? It is for the admin interface that doesn't exist.

bjoernricks commented:
> BTW I don't think httpd_basic_auth is actually ever invoked? It is for the admin interface that doesn't exist.

Yes it is. It's used in src/httpd_daap.c daap_response function line 800 for checking the authentication.

bjoernricks commented:
> I think that instead of inserting httpd_handle_cors_simple() in multiple locations, it would be better to change all occurances of evhttp_send_reply with httpd_send_reply, and then make the gzipping that httpd_send_reply does optional for the caller. Otherwise it becomes very inconsistent what headers forked-daapd returns.

As mentioned already I have some code for the preflight case ready. That code introduces a httpd_handle_cors_preflight function. It is a bit harder to decide if an options request must be answered. Imho it would not possible to do this at a central place for security reasons. E.g. it should be decided if the request contains a valid url before inspecting the request.

ejurgensen commented:
You're right about basic auth also being used for daap, I missed that.

> As mentioned already I have some code for the preflight case ready. That code introduces a httpd_handle_cors_preflight function. It is a bit harder to decide if an options request must be answered. Imho it would not possible to do this at a central place for security reasons. E.g. it should be decided if the request contains a valid url before inspecting the request.

Could you push a branch with the preflight code? Also, I'll admit that I have trouble understanding the last part. Could you elaborate on that?

bjoernricks commented:
Added code for the preflight case. Comments will follow tomorrow.

bjoernricks commented:
> >  It is a bit harder to decide if an options request must be answered. Imho it would not possible to do this at a central place for security reasons. E.g. it should be decided if the request contains a valid url before inspecting the request.
> 
> Could you push a branch with the preflight code? Also, I'll admit that I have trouble understanding the last part. Could you elaborate on that?

Of course! They easiest way to implement the preflight handling would be in httpd_gen_cb. That function is called for all requests to forked-daapd. But that would also mean it can only done before or after all other handling (e.g. all daap requests).

Handling it before going into the daap, dacp, rsp, ... code would imply to not being able to check for a valid url in the OPTIONS request. Then we also have to response to all OPTIONS request. Also for e.g. to OPTIONS /invalid-url.

Handling it in httpd_gen_cb after the daap, dacp, rcp, ... stuff would waste a lot of resources because it would process the OPTIONS request as it would be a GET currently.

In my patches I have added the preflight handling to the daap path AFTER the urls are checked. Imho that is safer without wasting resources.

bjoernricks commented:
@ejurgensen any news on this issue? I did start a web application for a daap client in the browser (https://github.com/bjoernricks/waap) in the meanwhile.

Also I did convert the daap protocol docu from http://tapjam.net/daap/ to markdown https://github.com/bjoernricks/daap-protocol Maybe you are willing to contribute. It's missing some up to date info.

ejurgensen commented:
Sorry for being so slow with this. I wanted to look closer at CORS to get a better understanding, but other things kept coming in the way. Hope to get it done soon. I also look forward to trying waap. You should consider adding a few screenshots. Is it mobile friendly?

Also good job with the documentation. Next time I am working on the daap parts, I will check to see if the documentation is up to date.

ejurgensen commented:
I finally got around to this issue, and now I have [made a branch](https://github.com/ejurgensen/forked-daapd/tree/cors), which I think should enable CORS headers. Haven't tested it very much yet, but perhaps you could also help with that?

Some things are different than in your PR:
1. CORS headers are enabled by default with "*" as origin. I think users should be able use waap out of the box.
2. I don't check for origin in normal requests. Checking every request for this seems like a waste of time :-)
3. I have not restricted which methods evhttp will accept - seemed a bit risky, since I am not sure what methods are actually used.

Notice that the first few commits of the branch don't have anything to do with CORS. I just happened to stumble on some possible cache/gzip performance improvements when I began looking at CORS.

bjoernricks commented:
Thanks so much for your efforts! I guess you understood 05426398e4e401b948ceffc2f8122b60b66caa88 as a restriction but it isn't. Currently option requests are answered with 501 by default. This commit allows option methods. When I remember correctly libevent only allows get and post by default. But haven't looked at it again yet.

bjoernricks commented:
Yes indeed. Not only get and post but no options. Found it in libevent/http.c:

```c
http_new_object(void)
{
	struct evhttp *http = NULL;

	if ((http = mm_calloc(1, sizeof(struct evhttp))) == NULL) {
		event_warn("%s: calloc", __func__);
		return (NULL);
	}

	http->timeout = -1;
	evhttp_set_max_headers_size(http, EV_SIZE_MAX);
	evhttp_set_max_body_size(http, EV_SIZE_MAX);
	evhttp_set_allowed_methods(http,
	    EVHTTP_REQ_GET |
	    EVHTTP_REQ_POST |
	    EVHTTP_REQ_HEAD |
	    EVHTTP_REQ_PUT |
	    EVHTTP_REQ_DELETE);

	TAILQ_INIT(&http->sockets);
	TAILQ_INIT(&http->callbacks);
	TAILQ_INIT(&http->connections);
	TAILQ_INIT(&http->virtualhosts);
	TAILQ_INIT(&http->aliases);

	return (http);
}
```

ejurgensen commented:
Yes, you are quite right. Think it should be fixed with the latest commit. I also included POST and HEAD, just in case.

ejurgensen commented:
Great, the branch is merged into master now
